### PR TITLE
4 step: Separate page for each news

### PR DESCRIPTION
**Pull request by EugeneTsalko**
Screenshot from 2023-10-16 23-11-28

### CODE REVIEW

File: src/app/article/ArticlePage.module.scss [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  margin-top: 5dvh;
+  padding: $spacing-xl;
+  border-radius: $border-radius;
+  box-shadow: 0px 0px 8px 0px rgba(0, 0, 0, 0.09);
+
+  &__image,
+  &__content {
+    width: 80dvw;
+
+    @include from-tablet {
```
Comment (ak-krivonosov) [inline src/app/article/ArticlePage.module.scss:19]: Please remove duplication

File: src/app/article/page.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
+'use client';
```
Comment (ak-krivonosov) [inline src/app/article/page.tsx:1]: If a user refreshes this page then there is an error appears. Please fix it as you see suitable

File: src/app/components/ReadMoreButton/ReadMoreButton.tsx [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import { NewsWithId } from '@/models/types';
+import { useRouter } from 'next/navigation';
+import styles from './ReadMoreButton.module.scss';
+
+type Props = {
+  article: NewsWithId;
+};
+
+function ReadMoreButton({ article }: Props) {
```
Comment (ak-krivonosov) [inline src/app/components/ReadMoreButton/ReadMoreButton.tsx]: Why do pass the whole article if you just need an id?

File: src/app/context.tsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
+'use client';
```
Comment (ak-krivonosov) [inline src/app/context.tsx]: Please create a context folder and put it there

Comment (ak-krivonosov) [reply src/app/context.tsx]: Also the file has a too generic name, please rename to  `news.context.tsx`

File: src/app/context.tsx [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+import React, { ReactNode, createContext, useContext, useState } from 'react';
+import { NewsWithId } from './models/types';
+
+type NewsState = {
```
Comment (ak-krivonosov) [inline src/app/context.tsx]: Please rename it to TNewsContext

File: src/app/context.tsx [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  context: NewsWithId[] | null;
```
Comment (ak-krivonosov) [inline src/app/context.tsx]: 1. Rename it to news
2. Arrays are quite convenient for types. Please remove `| null` and simply use `NewsWithId[]`

File: src/app/context.tsx [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  setContext(news: NewsWithId[]): void;
```
Comment (ak-krivonosov) [inline src/app/context.tsx]: Rename it to `setNews`

File: src/app/context.tsx [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+};
+
+const initialNewsState = {
+  context: null,
```
Comment (ak-krivonosov) [inline src/app/context.tsx]: Why do you use `null` as a default value? Would not it be better to use `[]`?

File: src/app/context.tsx [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+import React, { ReactNode, createContext, useContext, useState } from 'react';
+import { NewsWithId } from './models/types';
+
+type NewsState = {
+  context: NewsWithId[] | null;
+  setContext(news: NewsWithId[]): void;
+};
+
+const initialNewsState = {
```
Comment (ak-krivonosov) [inline src/app/context.tsx]: Why do you need this variable? You could put this object directly into `createContext`, couldn’t you?

File: src/app/context.tsx [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+  context: null,
+  setContext: () => {},
+};
+
+export const NewsContext = createContext<NewsState>(initialNewsState);
+
+export const useNewsContext = () => useContext(NewsContext);
+
+function Provider({ children }: { children: ReactNode }) {
+  const [context, setContext] = useState<NewsWithId[] | null>(null);
```
Comment (ak-krivonosov) [inline src/app/context.tsx]: `const [news, setNews] = useState<NewsWithId[]>([]);`

File: src/app/context.tsx [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+  // eslint-disable-next-line react/jsx-no-constructed-context-values
```
Comment (ak-krivonosov) [inline src/app/context.tsx]: Why do you use this comment? Isn’t it better to use `useMemo` instead of the comment?

File: src/app/page.tsx [hunk 12]
```diff
@@ -15,6 +16,7 @@ type Props = {
 function Home({ searchParams }: Props) {
   const [news, setNews] = useState<NewsWithId[] | null>(null);
```
Comment (ak-krivonosov) [inline src/app/page.tsx]: Use `[]` instead of the `null`

File: src/app/page.tsx [hunk 13]
```diff
@@ -34,6 +36,10 @@ function Home({ searchParams }: Props) {
     })();
   }, [searchParams.q]);
 
+  useEffect(() => {
+    setContext(news as NewsWithId[]);
```
Comment (ak-krivonosov) [inline src/app/page.tsx]: Please remove `as NewsWithId[]`

File: src/app/article/page.tsx [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+import getHumanDate from '@/utils/getHumanDate';
+import styles from './ArticlePage.module.scss';
+
+type Props = {
+  searchParams: { id: string };
+};
+
+function ArticlePage({ searchParams }: Props) {
+  const { context } = useNewsContext();
+  const article = context?.find((el) => el.id === searchParams.id) as NewsWithId;
```
Comment (ak-krivonosov) [inline src/app/article/page.tsx]: Please remove `as NewsWithId`. In this case when a user refreshes a page there is no context, so the user gets an error

File: src/app/article/page.tsx [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+  const humanDate = getHumanDate(article.publishedAt);
+
+  return (
+    <article className={styles.articlePage}>
+      {article.urlToImage && <img src={article.urlToImage} alt={article.title} className={styles.articlePage__image} />}
+      <h2>{article.title}</h2>
+      <div className={styles.articlePage__info}>
```
Comment (ak-krivonosov) [inline src/app/article/page.tsx]: Please use `<p>` here since you use it for text

**Review by ak-krivonosov — CHANGES_REQUESTED** [review]
Good luck

File: src/app/context.tsx (hunk 8 above)
Comment (EugeneTsalko) [reply src/app/context.tsx]: Null - search wasn't invoked, [ ] - search was invoked, but nothing was found.

File: src/app/article/page.tsx (hunk 14 above)
Comment (EugeneTsalko) [reply src/app/article/page.tsx]: Find can return NewsWithId or undefined. When user refreshes the page there is no context, and article is undefined, that's why error appears.

File: src/app/article/page.tsx (hunk 2 above)
Comment (EugeneTsalko) [reply src/app/article/page.tsx:1]: set last opened article in LocalStorage and get it if there is no context is ok?)

File: src/app/article/page.tsx [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+import getHumanDate from '@/utils/getHumanDate';
+import styles from './ArticlePage.module.scss';
+
+type Props = {
+  searchParams: { id: string };
+};
+
+function ArticlePage({ searchParams }: Props) {
+  const { news } = useNewsContext();
+  let article = news.find((el) => el.id === searchParams.id)!;
```
Comment (ak-krivonosov) [inline src/app/article/page.tsx]: You know TS was yelling at you for a reason before you added the `!`, if I try to open http://localhost:3000/article?id=12234 in a browser that doesn’t have anything in the localStorage, what will happen? There would be an error and your app is crashed. I can see 2 possible solutions here. First one is the easiest, if there is no article just redirect it to the home page. Second one is you can add a q parameter to the url so you’ll have something like that `[http://localhost:3000/?q=apple/article?id=12`](http://localhost:3000/?q=apple/article?id=12%60) and if there is no article and news you can make a request from a url and pick an article from it.

File: src/app/article/page.tsx (hunk 15 above)
Comment (ak-krivonosov) [reply src/app/article/page.tsx]: I’m sorry, I meant to use `<p>` instead of the `<div>`

File: src/app/context/news.context.tsx [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+};
+
+export const NewsContext = createContext<TNewsContext>({
+  news: [],
+  setNews: () => {},
+});
+
+export const useNewsContext = () => useContext(NewsContext);
+
+function Provider({ children }: { children: ReactNode }) {
```
Comment (ak-krivonosov) [inline src/app/context/news.context.tsx]: Please rename it to `NewsProvider`

File: src/app/page.tsx [hunk 18]
```diff
@@ -32,14 +32,14 @@ function Home({ searchParams }: Props) {
 
       setLoading(false);
     })();
-  }, [searchParams.q]);
+  }, [searchParams.q, setNews]);
```
Comment (ak-krivonosov) [inline src/app/page.tsx]: `[searchParams, setNews]`

**Review by ak-krivonosov — CHANGES_REQUESTED** [review]
Good luck

File: src/app/article/page.tsx (hunk 15 above)
Comment (ak-krivonosov) [reply src/app/article/page.tsx]: `<p className={styles.articlePage__info}>`

**Review by ak-krivonosov — APPROVED** [review]
Fix the comment and you can merge

**Pull request merged**